include/rsp_dma.inc

```diff
@@ -0,0 +1,183 @@
+########################################################
+# Include this file wherever you prefer in your text segment
+########################################################
+
+########################################################
+# DMA functions:
+#   DMAIn / DMAInAsync / DMAOut / DMAOutAsync / DMAExec
+#
+# These functions are used to issue DMA transfers between
+# RDRAM and internal DMEM:
+#
+#   * DMAIn: run a RDRAM -> DMEM transfer, waiting for completion
+#   * DMAInAsync: start a background RDRAM -> DMEM transfer
+#   * DMAOut: run a DMEM -> RDRAM transfer, waiting for completion
+#   * DMAOutAsync: start a background DMEM -> RDRAM transfer
+#   * DMAExec: single entry point for any type of transfer. It can be used
+#     to write direction-agnostic transfer code that can be then invoked
+#     passing in the direction in a register.
+#
+# The RSP DMA controller allows to enqueue one transfer while another one
+# is in progress. You can use the *Async functions to do this. If you try
+# to enqueue a third transfer, though, the *Async function will automatically
+# block waiting for the first one to finish. This allows the code to call
+# *Async functions freely and get "best effort" background transfers.
+# 
+# When you need to explicitly synchronize (eg: to access read data), you can
+# use DMAWaitReady to wait until the DMA engine is ready to receive another
+# transfer, or DMAWaitIdle to wait until all DMA transfers are finished. See
+# their documentation below.
+#
+# The DMA controller is able to read/write "rectangular portions"
+# of memory in RDRAM (that is, you can define width, height and
+# pitch to extract a portion of a larger image). Accesses in DMEM
+# are always linear, though. If you don't need this feature,
+# you can set the transfer size as width and use 1 as height: this way,
+# the pitch will be ignored.
+#
+# At the hardware level, both the DMEM and the RDRAM addresses are
+# treated as 8-byte aligned (that is, the lowest 3 bits are ignored).
+# If you issue a transfer from RDRAM with an unaligned address (in s0),
+# the DMEM address (s4) will be updated to reflect where the byte that
+# was actually requested ended up. For instance, if s0 = 0x80041203, s4=0x100,
+# after the function s4 will be 0x103, as that is the address where the
+# byte at 0x80041203 ended up being written to.
+#
+# INPUT:
```

Comment:
Right, I forgot to discuss about this.

In general, all my microcode does not follow the C MIPS ABI. That's because the RSP context is very different. First of all, all code is assembly; second, there is often no stack, so that also makes it impossible to both fully follow the ABI, and to do something as simple as saving/restore registers; third, the MIPS ABI forces you to shuffle registers a lot (saving them on the stack) because it has fixed assignments.

The MIPS ABI (or any ABI, for that matter) make sense in the context of allowing foreign code to interoperate, as it is normally the case for C which is compiled in different translation units so you don't know what the other code is doing. If you were to compile the 100% of a binary from a single C file (including standard libraries, etc.) the compiler could in theory forego the ABI and find an optimal allocation for the specific problem at hand -- no compiler implements this because it is such a specialistic use case that nobody needs.

And in fact, I think we also don't need it for RSP. RSP ucode can be written using some global register allocation where you have a few "local, scratch registers" that you can use freely and can be freely clobbered, and most registers are simply "globally allocated" for large sections of the code, or even the whole microcode. If you look at SGI RSP code snippets, they don't follow the standard MIPS ABI as well.

So what I should do, instead, is to document the convention I came up with and that all my microcode follows; given that I'm going to submit that code and that any convention is good, I claim the right of first arrival here so that we can use mine without me having to rename registers in all my code (which also would be troublesome to do without introducing bugs).

My convention is:

 * microcode always starts at 0x0000 in IMEM
 * `set noreorder` by default, to squeeze every last ounce of performance
 * `set at` by default, but you can switch to `set noat` to have one more scratch register when needed (eg: very useful in macro and helpers).
 * 2 levels of nested function calls are supported; use `sp` as `ra2` to implement the two levels. (Ab)use tail calls to help with this. If you need more, try to refactor, or in the worse case, you need to allocate another register as `ra3`.
 * `t0-t3`: scratch value registers, often used also as arguments to functions. These are considered clobbered by defaults in function calls.
 * `s0-s4`: scratch pointer registers, often used also as arguments to functions. These are often clobbered by functions calls.
 * `a0-a3`: display list input to be processed (eg: the 64bit primitive being processed can go in `a0/a1`). This applies if you process some sort of display list input of course.
 * anything else: use it for keeping values around in larger sections of the code (or even the whole microcode). Rename each register with #define to begin using, and then #undef after you don't need it. Wish gas had proper register renaming, but it doesn't, so we'll use preprocessor for this.

Reply:
I didn't mean strictly full ABI but overall conventions, this is all OK for me. Would be nice to write this somewhere though.

Reply:
Done.